### PR TITLE
rpc: Improve signrawtransaction error reporting

### LINKED ISSUE #13218: signrawtransactionwithkey get me confusing error message even when hash160(redeemScript) != scriptPubKey.
signrawtransactionwithkey get me confusing error message even when hash160(redeemScript) != scriptPubKey.
Obviously `hash 160 (redeemScript)! = ScriptPubKey`
If so, why do not give me a better message?

```
signrawtransactionwithkey  "0200000001cfe62b9f1c52bc2998f9a0e6d6ee35430c93ec653d5691935178147a43090fee0000000000ffffffff01f0afd320766191d4c5c088ac00000000" "[\"cSXHtKRXWnCLWhuie8WC13cKF3qdnKL8vG3BiRpxgCXZkK2CHe9u\"]"  "[{\"txid\":\"ee0f09437a1478519391563d65ec930c4335eed6e6a0f99829bc521c9f2be6cf\",\"vout\":0,\"scriptPubKey\":\"a914111111111111111111111111111111111111111187\",\"redeemScript\":\"5221033d9aecbced8e776c03d283f6c0d64c6a301ad142d74d500fe8860c9b362fa4932102a4df3cb4e5286e180b7c68e323c778e32350a604d4f4a74c290d1c60c721687a52ae\"}]"
{
  "hex": "0200000001cfe62b9f1c52bc2998f9a0e6d6ee35430c93ec653d5691935178147a43090fee0000000000ffffffff01f0a29a3b000000001976a9146eeff5b6a0960be69ad0941fd320766191d4c5c088ac00000000",
  "complete": false,
  "errors": [
    {
      "txid": "ee0f09437a1478519391563d65ec930c4335eed6e6a0f99829bc521c9f2be6cf",
      "vout": 0,
      "witness": [
      ],
      "scriptSig": "",
      "sequence": 4294967295,
      **"error": "Unable to sign input, invalid stack size (possibly missing key)"**
    }
  ]
}

signrawtransactionwithkey  "0200000001cfe62b9f1c52bc2998f9a0e6d6ee35430c93ec653d5691935178147a43090fee0000000000ffffffff01f0afd320766191d4c5c088ac00000000" "[\"cSXHtKRXWnCLWhuie8WC13cKF3qdnKL8vG3BiRpxgCXZkK2CHe9u\"]"  "[{\"txid\":\"ee0f09437a1478519391563d65ec930c4335eed6e6a0f99829bc521c9f2be6cf\",\"vout\":0,\"scriptPubKey\":\"a914648a73e6a0db686b547a66b7b38f3f7af61f50ae87\",\"redeemScript\":\"52211111111111111111111111111111111111111111111111111111111111111111112102a4df3cb4e5286e180b7c68e323c778e32350a604d4f4a74c290d1c60c721687a52ae\"}]"
{
  "hex": "0200000001cfe62b9f1c52bc2998f9a0e6d6ee35430c93ec653d5691935178147a43090fee0000000000ffffffff01f0a29a3b000000001976a9146eeff5b6a0960be69ad0941fd320766191d4c5c088ac00000000",
  "complete": false,
  "errors": [
    {
      "txid": "ee0f09437a1478519391563d65ec930c4335eed6e6a0f99829bc521c9f2be6cf",
      "vout": 0,
      "witness": [
      ],
      "scriptSig": "",
      "sequence": 4294967295,
      **"error": "Unable to sign input, invalid stack size (possibly missing key)"**
    }
  ]
}

signrawtransactionwithkey  "0200000001cfe62b9f1c52bc2998f9a0e6d6ee35430c93ec653d5691935178147a43090fee0000000000ffffffff01f0afd320766191d4c5c088ac00000000" "[\"cSXHtKRXWnCLWhuie8WC13cKF3qdnKL8vG3BiRpxgCXZkK2CHe9u\"]"  "[{\"txid\":\"ee0f09437a1478519391563d65ec930c4335eed6e6a0f99829bc521c9f2be6cf\",\"vout\":0,\"scriptPubKey\":\"a914648a73e6a0db686b547a66b7b38f3f7af61f50ae87\",\"redeemScript\":\"5221033d9aecbced8e776c03d283f6c0d64c6a301ad142d74d500fe8860c9b362fa4932111111111111111111111111111111111111111111111111111111111111111111152ae\"}]"
{
  "hex": "0200000001cfe62b9f1c52bc2998f9a0e6d6ee35430c93ec653d5691935178147a43090fee0000000000ffffffff01f0a29a3b000000001976a9146eeff5b6a0960be69ad0941fd320766191d4c5c088ac00000000",
  "complete": false,
  "errors": [
    {
      "txid": "ee0f09437a1478519391563d65ec930c4335eed6e6a0f99829bc521c9f2be6cf",
      "vout": 0,
      "witness": [
      ],
      "scriptSig": "",
      "sequence": 4294967295,
      "error": "Unable to sign input, invalid stack size (possibly missing key)"
    }
  ]
}

```
<What behavior did you expect?>
"error": "Unable to sign input, the scriptPubKey does not match redeemScript”

<What version of Bitcoin Core are you using, where did you get it (website, self-compiled, etc)? >
```
./bitcoind --version
Bitcoin Core Daemon version v0.16.99.0-81d0d565a
Copyright (C) 2009-2018 The Bitcoin Core developers

```

### LINKED ISSUE #14823: Confusing output when partial signing multisig transaction
Unexpected output during multisig signing. See the following example when signing a 2/2 key.

```
bc signrawtransactionwithkey 0200000001b8a7873d892de2930cdb00a6c69dadbe8d553e2bef0688186a71009fd82062460100000000ffffffff028093dc140000000017a914a732060d1955ab93c50c1096597e2ea466b9937d8780d6e34c0000000017a9142d2aad185524b244ef2b2f6a0507d6f187d02fbe8700000000 '["cU3C1Uu4oN8ZPXrCz55GyjmrsaeoF8UhmBai2d9jzq9iw7ty9EMc"]' '[{"txid":"466220d89f00716a188806ef2b3e558dbead9dc6a600db0c93e22d893d87a7b8", "vout":1,"scriptPubKey":"a9142d2aad185524b244ef2b2f6a0507d6f187d02fbe87","redeemScript":"5221032559b7054ff5468aa9508579379b62c5f6b31145e5bfe743fc66c0ec72608dab21030eaf6ea71e09bd6f6b10a3ee42cc1a445d9020627a4569f17b7fbf1d5801959a52ae"}]'
```

```
{
  "hex": "0200000001b8a7873d892de2930cdb00a6c69dadbe8d553e2bef0688186a71009fd820624601000000920047304402202d3ccf2360f990235e2e0ce17f5f28f4313f413301869b554c4e79905de93508022009a4acd673571d7924e8c0cd192386972a72316e8195338ebb8577a2a2941ed70100475221032559b7054ff5468aa9508579379b62c5f6b31145e5bfe743fc66c0ec72608dab21030eaf6ea71e09bd6f6b10a3ee42cc1a445d9020627a4569f17b7fbf1d5801959a52aeffffffff028093dc140000000017a914a732060d1955ab93c50c1096597e2ea466b9937d8780d6e34c0000000017a9142d2aad185524b244ef2b2f6a0507d6f187d02fbe8700000000",
  "complete": false,
  "errors": [
    {
      "txid": "466220d89f00716a188806ef2b3e558dbead9dc6a600db0c93e22d893d87a7b8",
      "vout": 1,
      "witness": [
      ],
      "scriptSig": "0047304402202d3ccf2360f990235e2e0ce17f5f28f4313f413301869b554c4e79905de93508022009a4acd673571d7924e8c0cd192386972a72316e8195338ebb8577a2a2941ed70100475221032559b7054ff5468aa9508579379b62c5f6b31145e5bfe743fc66c0ec72608dab21030eaf6ea71e09bd6f6b10a3ee42cc1a445d9020627a4569f17b7fbf1d5801959a52ae",
      "sequence": 4294967295,
      "error": "Signature must be zero for failed CHECK(MULTI)SIG operation"
    }
  ]
}
```

```
bc signrawtransactionwithkey 0200000001b8a7873d892de2930cdb00a6c69dadbe8d553e2bef0688186a71009fd820624601000000920047304402202d3ccf2360f990235e2e0ce17f5f28f4313f413301869b554c4e79905de93508022009a4acd673571d7924e8c0cd192386972a72316e8195338ebb8577a2a2941ed70100475221032559b7054ff5468aa9508579379b62c5f6b31145e5bfe743fc66c0ec72608dab21030eaf6ea71e09bd6f6b10a3ee42cc1a445d9020627a4569f17b7fbf1d5801959a52aeffffffff028093dc140000000017a914a732060d1955ab93c50c1096597e2ea466b9937d8780d6e34c0000000017a9142d2aad185524b244ef2b2f6a0507d6f187d02fbe8700000000 '["cSB1o7v1QBZDHJropQqBYfuiUA4Js1o24uzkgB2vYeND1st5uGqV"]' 
```

```
{
  "hex": "0200000001b8a7873d892de2930cdb00a6c69dadbe8d553e2bef0688186a71009fd820624601000000d90047304402202d3ccf2360f990235e2e0ce17f5f28f4313f413301869b554c4e79905de93508022009a4acd673571d7924e8c0cd192386972a72316e8195338ebb8577a2a2941ed70147304402206fe3aa79e02894ec40d3f6e9df7d7056c5b50cd7a39e1cde3aa3dd097c76798f0220464cd0a39690cbf28687b3b695eb0dc59ca73aef624b16b9586ade31a0761b3101475221032559b7054ff5468aa9508579379b62c5f6b31145e5bfe743fc66c0ec72608dab21030eaf6ea71e09bd6f6b10a3ee42cc1a445d9020627a4569f17b7fbf1d5801959a52aeffffffff028093dc140000000017a914a732060d1955ab93c50c1096597e2ea466b9937d8780d6e34c0000000017a9142d2aad185524b244ef2b2f6a0507d6f187d02fbe8700000000",
  "complete": true
}
```

When dealing with the RPC calls alone, it was difficult to realize that the partial signing was actually working. This is due to the this specific portion of the response:

```
  "errors": [
    {
      "txid": "466220d89f00716a188806ef2b3e558dbead9dc6a600db0c93e22d893d87a7b8",
      "vout": 1,
      "witness": [
      ],
      "scriptSig": "0047304402202d3ccf2360f990235e2e0ce17f5f28f4313f413301869b554c4e79905de93508022009a4acd673571d7924e8c0cd192386972a72316e8195338ebb8577a2a2941ed70100475221032559b7054ff5468aa9508579379b62c5f6b31145e5bfe743fc66c0ec72608dab21030eaf6ea71e09bd6f6b10a3ee42cc1a445d9020627a4569f17b7fbf1d5801959a52ae",
      "sequence": 4294967295,
      "error": "Signature must be zero for failed CHECK(MULTI)SIG operation"
    }

```

Based on the RPC respons on here: https://bitcoin.org/en/developer-examples#p2sh-multisig

I expected the successful partial signature to be as follows:
```
{
    "hex" : "010000000175e1769813db8418fea17576694af1ff31cb2b512\
             b7333e6eb42f030d0d7787200000000b5004830450221008d5e\
             c57d362ff6ef6602e4e756ef1bdeee12bd5c5c72697ef1455b3\
             79c90531002202ef3ea04dfbeda043395e5bc701e4878c15baa\
             b9c6ba5808eb3d04c91f641a0c014c69522103310188e911026\
             cf18c3ce274e0ebb5f95b007f230d8cb7d09879d96dbeab1aff\
             210243930746e6ed6552e03359db521b088134652905bd2d154\
             1fa9124303a41e95621029e03a901b85534ff1e92c43c74431f\
             7ce72046060fcf7a95c37e148f78c7725553aeffffffff01c0b\
             c973b000000001976a914b6f64f5bf3e38f25ead28817df7929\
             c06fe847ee88ac00000000",
    "complete" : false
}
```

I'm using `Bitcoin Core RPC client version v0.17.0`

I suspect the issue is related to this: https://github.com/bitcoin/bitcoin/issues/9988, however they're using a different RPC call.

### PR DESCRIPTION
**Pull request by ajtowns**
Two fixes for `signrawtransactionwith{key,wallet}` (in addition to #16250): one that checks redeemScript/witnessScript matches scriptPubKey (and if both are provided that they match each other sanely), and the other changes the warning when some-but-not-all the signatures for a CHECKMULTISIG are provided to something that suggests more signatures may be all that's required.

Fixes: #13218
Fixes: #14823

### CODE REVIEW

**Comment by ajtowns** [comment]
With this patch, partial signature error message looks like:

    "error": "CHECK(MULTI)SIG failing with non-zero signature (possibly need more signatures)"

instead of:

    "error": "Signature must be zero for failed CHECK(MULTI)SIG operation"

Mismatching scripts give a new error of either:

    error code: -3
    error message:
    redeemScript/witnessScript does not match scriptPubKey

or

    error code: -3
    error message:
    redeemScript does not correspond to witnessScript

instead of an apparently successful operation (but unchanged rawtx) with embedded error of:

    "error": "Witness program was passed an empty witness"

File: src/rpc/rawtransaction_util.cpp [hunk 1]
```diff
@@ -221,6 +221,9 @@ UniValue SignTransaction(CMutableTransaction& mtx, const UniValue& prevTxsUnival
                     // Automatically also add the P2WSH wrapped version of the script (to deal with P2SH-P2WSH).
                     keystore->AddCScript(GetScriptForWitness(witnessScript));
                 }
+                if (rs.isNull() && ws.isNull()) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Missing redeemScript/witnessScript");
+                }
```
Comment (MarcoFalke) [inline src/rpc/rawtransaction_util.cpp]: I feel like this has already been merged

Comment (ajtowns) [reply src/rpc/rawtransaction_util.cpp]: Yeah, I just rebased it (` ajtowns force-pushed the ajtowns:201906-signrawerror branch from 073a2f8 to f7e4676`) but github's not picking up the new HEAD?

Comment (MarcoFalke) [reply src/rpc/rawtransaction_util.cpp]: It is well known for GitHub to show outdated branches in the web view (see also https://github.com/bitcoin/bitcoin/issues/15847#issuecomment-484871831)


I guess you have to force push twice?

Comment (ajtowns) [reply src/rpc/rawtransaction_util.cpp]: Looks like it's doing the right thing now (I pushed a dummy commit and force-pushed it out of existence). You should be seeing f7e4676591c9027a6fe4dd1ebc9e1ed35b27574f as HEAD.

**Comment by meshcollider** [comment]
Concept ACK, looks right to me after an initial read

File: src/rpc/rawtransaction_util.cpp [hunk 2]
```diff
@@ -196,32 +196,73 @@ void ParsePrevouts(const UniValue& prevTxsUnival, FillableSigningProvider* keyst
...
+                        // either script, so we got unusable parameters
+                        throw JSONRPCError(RPC_INVALID_PARAMETER, "redeemScript/witnessScript does not match scriptPubKey");
+                    }
+                } else if (is_p2wsh) {
+                    const CTxDestination p2wsh{WitnessV0ScriptHash(script)};
+                    // plain p2wsh; could throw an error if script
+                    // was specified by redeemScript rather than
+                    // witnessScript (ie, ws.IsNull() == true), but
+                    // accept it for backwards compat
+
```
Comment (instagibbs) [inline src/rpc/rawtransaction_util.cpp]: nit: extra newline

File: src/rpc/rawtransaction_util.cpp [hunk 3]
```diff
@@ -196,32 +196,73 @@ void ParsePrevouts(const UniValue& prevTxsUnival, FillableSigningProvider* keyst
...
                     throw JSONRPCError(RPC_INVALID_PARAMETER, "Missing redeemScript/witnessScript");
                 }
+
+                // work from witnessScript when possible
+                std::vector<unsigned char> scriptData(!ws.isNull() ? ParseHexV(ws, "witnessScript") : ParseHexV(rs, "redeemScript"));
+                CScript script(scriptData.begin(), scriptData.end());
+                keystore->AddCScript(script);
+                // Automatically also add the P2WSH wrapped version of the script (to deal with P2SH-P2WSH).
+                // This is done for redeemScript only for compatibility, it is encouraged to use the explicit witnessScript field instead.
+                CScript witnessScript{GetScriptForWitness(script)};
```
Comment (instagibbs) [inline src/rpc/rawtransaction_util.cpp]: this is the witness output script aka script with version number and witness program, not `witnessScript`, right?

Comment (meshcollider) [reply src/rpc/rawtransaction_util.cpp]: Yes and this is also not correct I think? (The current behavior too). GetScriptForWitness will return a P2WSH scriptPubKey script, it won't wrap it in P2SH, so we'll end up with things like P2WSH-P2SH I think. I noticed this in #16841

Comment (meshcollider) [reply src/rpc/rawtransaction_util.cpp]: Yes, it is technically a P2SH redeemScript in the form of a P2WSH scriptPubKey

File: src/rpc/rawtransaction_util.cpp (hunk 2 above)
Comment (ajtowns) [reply src/rpc/rawtransaction_util.cpp]: wow, that is nitty! fixed

File: src/rpc/rawtransaction_util.cpp (hunk 3 above)
Comment (ajtowns) [reply src/rpc/rawtransaction_util.cpp]: fixed

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/16251/commits/ec4c79326bb670c2cc1757ecfb1900f8460c5257

**Review by meshcollider — COMMENTED** [review]
utACK ec4c79326bb670c2cc1757ecfb1900f8460c5257

I really like this change

**Comment by achow101** [comment]
Code Review ACK ec4c79326bb670c2cc1757ecfb1900f8460c5257

**Pull request merged**